### DISCUSSION TITLE
자바스크립트 엔진이 실행 중에 코드를 재컴파일하는 이유는 무엇인가요?

### DISCUSSION THREAD

**Discussion by lledellebell** [diving-javascript/you-dont-know-js-yet#58] [1. 스코프] [answered]
<details>
<summary>관련 챕터</summary>
1.3 코드 컴파일 (Compiling Code)
</details>

**Comment by wogus216** [comment, marked as answer]
<details>
<summary>답변</summary>



재컴파일은 JIT(Just-In-Time) 컴파일러과 관련된 개념이고, JIT 컴파일러는 코드를 실행하는 동안 실시간으로 컴파일하고 최적화하여 성능을 향상시키는 역할을 합니다. JIT 컴파일링의 주요 특징 중 하나는 런타임에 코드를 재컴파일하고 최적화하는 능력

**JIT 컴파일러 주요개념**

1. 초기 컴파일: 프로그램이 처음 실행될 때, JIT 컴파일러는 코드를 빠르게 컴파일하여 실행 할 수 있게하고 다만, 이때 코드는 최적화가 안된 상태
2. 프로 파일링: 코드가 실행되는 동안 JIT 컴파일러는 프로그램의 실행 패턴과 데이터 타입을 분석, 이정보를 바탕으로 코드의 성능을 최적화할 기회를 찾음
3. 최적화 컴파일: 충분한 프로파일링 정보가 수집되면, JIT 컴파일러는 코드를 재컴파일하여 최적화된 기계어 코드를 생성
4. 디옵티마이제이션(최적화 해제): 최적화된 코드가 예상치 못한 데이터 타입이나 실행 경로를 만날 수 있습니다. 이 경우 JIT 컴파일러는 최적화된 코드를 디옵티마이제이션하여 안정적인 실행을 보장

</details>
